### PR TITLE
added support for unseq, par_unseq for hpx::make_heap algorithm

### PR DESCRIPTION
**Pull request by Johan511**
Added unseq and par_unseq to hpx::make_heap algorithm

par_unseq => existing function calls sift_down_range in parallel. Overloaded sift_down_range function to call appropriate function based on value of hpx::is_unsequenced_execution_policy<ExPolicy>

unseq => current sequential implementation calls sequential_make_heap(), overloaded it to call appropriate function for seq vs unseq.

Should we consider using sift_down_range for sequential implementation too?

### CODE REVIEW

File: libs/core/algorithms/include/hpx/parallel/algorithms/make_heap.hpp [hunk 1]
```diff
@@ -293,6 +293,33 @@ namespace hpx::parallel {
             }
         }
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/make_heap.hpp:294]: Is the original implementation of `sift_down_range()` still needed?

File: libs/core/algorithms/include/hpx/parallel/algorithms/make_heap.hpp [hunk 2]
```diff
@@ -293,6 +293,33 @@ namespace hpx::parallel {
...
+            HPX_IVDEP HPX_UNROLL HPX_VECTORIZE
+            for (std::size_t i = 0; i != count; ++i)
+            {
+                sift_down(first, comp, proj, len, start - i);
+            }
+            // clang-format on
+        }
+
         template <typename Iter, typename Sent, typename Comp, typename Proj>
         constexpr Iter sequential_make_heap(
```
Comment (hkaiser) [inline libs/core/algorithms/include/hpx/parallel/algorithms/make_heap.hpp:324]: Same here, is this still needed?

**Review by hkaiser — CHANGES_REQUESTED** [review]
Have you done some performance analysis to make sure the changes have the desired effect?

File: libs/core/algorithms/include/hpx/parallel/algorithms/make_heap.hpp (hunk 1 above)
Comment (Johan511) [reply libs/core/algorithms/include/hpx/parallel/algorithms/make_heap.hpp:294]: I don't think so, the only place where sift_down_range() is called has been overloaded. We can delete the original implementation.

File: libs/core/algorithms/include/hpx/parallel/algorithms/make_heap.hpp (hunk 2 above)
Comment (Johan511) [reply libs/core/algorithms/include/hpx/parallel/algorithms/make_heap.hpp:324]: I don't think so, the only place where sequential_make_heap() is called has been overloaded. We can delete the original implementation.

**Comment by Johan511** [comment]
Haven't done any performance analysis as of yet, but have confirmed (using gdb) that the desired overload is begin called.
Any advice on how formally do a performance analysis?

**Comment by Johan511** [comment]
Following are the performance test graphs


hpx::make_heap(hpx::execution::par, vec1.begin(), vec1.end());
par


hpx::make_heap(hpx::execution::par_unseq, vec1.begin(), vec1.end());
par_unseq

**Review by hkaiser — COMMENTED** [review]
From your graphs, your patch has no noticeable effect on the algorithm's performance. Are you sure the compiler is able to see through the `sift_down` function to vectorize the loop?

**Comment by Johan511** [comment]
Yes, I have gone through the assembly generated, sift down is not being vectorized. I have talked to few people with more experience than me in this field and they believe the function is too complex and might not be vectorized, they have advised me to look into std::experimental::simd and also into breaking the loop into parts and vectorize a simpler part of it. I will try doing it and benchmarking them. It is probably better to close this issue as the PR is unlikely to work.

**Comment by hkaiser** [comment]
> Yes, I have gone through the assembly generated, sift down is not being vectorized. I have talked to few people with more experience than me in this field and they believe the function is too complex and might not be vectorized, they have advised me to look into std::experimental::simd and also into breaking the loop into parts and vectorize a simpler part of it. I will try doing it and benchmarking them. It is probably better to close this issue as the PR is unlikely to work.

I'd be happy with leaving this PR open. You can always (force-) push more changes here to modify the patch.

**Comment by hkaiser** [comment]
@Johan511 could you please rebase this onto master, now that the release is out?

**Pull request closed without merging**